astroplan/scheduling.py

```diff
@@ -124,8 +126,170 @@ def components(self, val):
+
+    def new_slots(self, slot_index, start_time, end_time):
+        # this is intended to be used such that there aren't consecutive unoccupied slots
+        new_slots = self.slots[slot_index].split_slot(start_time, end_time)
+        return new_slots
+
+    def insert_slot(self, slot_index, start_time, block):
+        if block.duration + self.slew_duration > self.slots[slot_index].duration:
+            raise ValueError('constraint application failed and did not leave enough time')
+        if isinstance(block, ObservingBlock):
+            tb = TransitionBlock.from_duration(self.slew_duration)
+            if self.slots[slot_index].end-(start_time+block.duration) < tb.duration:
+                self.insert_slot(slot_index, self.slots[slot_index].end-tb.duration, tb)
+                start_time = self.slots[slot_index].end - block.duration
+            else:
+                self.insert_slot(slot_index, start_time+block.duration, tb)
+            # TODO: make it shift observing/transition blocks to fill small amounts of open space
+            block.end_time = start_time+block.duration
+        earlier_slots = self.slots[:slot_index]
+        later_slots = self.slots[slot_index+1:]
+        end_time = start_time+block.duration
+        block.start_time = start_time
+        new_slots = self.new_slots(slot_index, start_time, end_time)
+        for new_slot in new_slots:
+            if new_slot.middle:
+                new_slot.occupied = True
+                if isinstance(block, ObservingBlock):
+                    new_slot.OB = block
+                elif isinstance(block, TransitionBlock):
+                    new_slot.TB = block
+        self.slots = earlier_slots + new_slots + later_slots
+        return earlier_slots + new_slots + later_slots
+    
+    def remove_slot(self, slot_index, start_time, end_time):
+        earlier_slots = self.slots[:slot_index]
+        later_slots = self.slots[slot_index+1:]
+        new_slots = self.new_slots(slot_index, start_time, end_time)
+        for new_slot in new_slots:
+            if new_slot.middle:
+                new_slots.remove(new_slot)
+        return earlier_slots + new_slots + later_slots
+    
+    @classmethod
+    def from_constraints(cls, start_time, end_time, constraints):
+        sch = cls(start_time, end_time, constraints=constraints)
+        sch.apply_constraints()
+        return sch
+
+
+class Slot(object):
+    """
+    A time slot within the schedule
+    
+    Parameters:
+    -----------
+    start_time : `~astropy.time.Time`
+        The starting time of the slot
+    end_time : `~astropy.time.Time`
+        The ending time of the slot    
+    """
+    
+    def __init__(self, start_time, end_time):
+        self.start = start_time
+        self.end = end_time
+        self.duration = end_time-start_time
+        self.occupied = False
+        self.middle = False
+        self.OB = False
+        self.TB = False
```

Comment:
Shouldn't a slot contain the block instance? For example, a slot filled with an OB will contain the `ObservingBlock` object, on an attribute called `self.block` or something.

Then you won't need the `Slot.OB` or `Slot.TB` attributes, you'd be able to check if `isinstance(self.block, ObservingBlock)`, for instance.